Game/Source/GDKTestGyms/AUptimeGameMode.cpp

```diff
@@ -2,482 +2,159 @@
+void AUptimeGameMode::StartCustomNPCSpawning()
 {
-	if (NumPoints <= 0)
-	{
-		UE_LOG(LogUptimeGymGameMode, Warning, TEXT("Generating grid settings with non-postive number of points (%d)"), NumPoints);
-		OutNumRows = 0;
-		OutNumCols = 0;
-		OutMinRelativeX = 0;
-		OutMinRelativeY = 0;
-		return;
-	}
-
-	OutNumRows = FMath::RoundToInt(FMath::Sqrt(NumPoints));
-	OutNumCols = FMath::CeilToInt(NumPoints / static_cast<float>(OutNumRows));
-	const int GridWidth = (OutNumCols - 1) * DistBetweenPoints;
-	const int GridHeight = (OutNumRows - 1) * DistBetweenPoints;
-	OutMinRelativeX = FMath::RoundToInt(-GridWidth / 2.0);
-	OutMinRelativeY = FMath::RoundToInt(-GridHeight / 2.0);
+	Super::StartCustomNPCSpawning();
+	SpawnCrossServerActors();
 }
 
-void AUptimeGameMode::GenerateSpawnPoints(int CenterX, int CenterY, int SpawnPointsNum)
+void AUptimeGameMode::SpawnCrossServerActors()
 {
-	// Spawn in the air above terrain obstacles (Unreal units).
-	const int Z = 300;
-
-	const int DistBetweenSpawnPoints = 300; // In Unreal units.
-	int NumRows, NumCols, MinRelativeX, MinRelativeY;
-	GenerateGridSettings(DistBetweenSpawnPoints, SpawnPointsNum, NumRows, NumCols, MinRelativeX, MinRelativeY);
-
-	UWorld* World = GetWorld();
+	UWorld* const World = GetWorld();
 	if (World == nullptr)
 	{
-		UE_LOG(LogUptimeGymGameMode, Error, TEXT("Cannot spawn spawnpoints, world is null"));
+		UE_LOG(LogUptimeGymGameMode, Error, TEXT("Error spawning, World is null"));
 		return;
 	}
 
-	for (int i = 0; i < SpawnPointsNum; ++i)
-	{
-		const int Row = i % NumRows;
-		const int Col = i / NumRows;
-
-		const int X = CenterX + MinRelativeX + Col * DistBetweenSpawnPoints;
-		const int Y = CenterY + MinRelativeY + Row * DistBetweenSpawnPoints;
-
-		FActorSpawnParameters SpawnInfo{};
-		SpawnInfo.Owner = this;
-		SpawnInfo.Instigator = NULL;
-		SpawnInfo.bDeferConstruction = false;
-		SpawnInfo.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
-
-		const FVector SpawnLocation = FVector(X, Y, Z);
-		UE_LOG(LogUptimeGymGameMode, Log, TEXT("Creating a new PlayerStart at location %s."), *SpawnLocation.ToString());
-		SpawnPoints.Add(World->SpawnActor<APlayerStart>(APlayerStart::StaticClass(), SpawnLocation, FRotator::ZeroRotator, SpawnInfo));
-	}
-}
-
-void AUptimeGameMode::SpawnNPCs(int NumNPCs)
-{
-	NPCSToSpawn = NumNPCs;
-}
-
-void AUptimeGameMode::SpawnNPC(const FVector& SpawnLocation, const FBlackboardValues& BlackboardValues)
-{
-	UWorld* const World = GetWorld();
-	if (World == nullptr)
+	if (SpawnManager == nullptr)
 	{
-		UE_LOG(LogUptimeGymGameMode, Error, TEXT("Error spawning NPC, World is null"));
+		UE_LOG(LogUptimeGymGameMode, Error, TEXT("Error spawning, SpawnManager is null"));
 		return;
 	}
 
-	if (NPCSpawner == nullptr)
-	{
-		ABenchmarkGymNPCSpawner* Spawner = nullptr;
-		for (TActorIterator<ABenchmarkGymNPCSpawner> It(GetWorld()); It; ++It)
+	SpawnManager->ForEachZoneArea([this, World](USpawnArea& ZoneArea) {
```

Comment:
nit - could make this `ForEachArea` and have the user pass in an enum to match the type of area they want to execute on